router/data_store.go

```diff
@@ -89,38 +98,224 @@ INSERT INTO ` + tableNameTCP + ` (parent_ref, service, leader, port)
-			sqlAddRouteHTTP,
-			r.ParentRef,
-			r.Service,
-			r.Leader,
-			r.Domain,
-			r.TLSCert,
-			r.TLSKey,
-			r.Sticky,
-			r.Path,
-		).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
+		err = d.addHTTP(r)
 	case tableNameTCP:
-		err = d.pgx.QueryRow(
-			sqlAddRouteTCP,
-			r.ParentRef,
-			r.Service,
-			r.Leader,
-			r.Port,
-		).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
+		err = d.addTCP(r)
 	}
 	r.Type = d.routeType
-	if postgres.IsUniquenessError(err, "") {
-		err = ErrConflict
-	} else if postgres.IsPostgresCode(err, postgres.RaiseException) {
-		err = ErrInvalid
+	if err != nil {
+		if postgres.IsUniquenessError(err, "") {
+			err = ErrConflict
+		} else if postgres.IsPostgresCode(err, postgres.RaiseException) {
+			err = ErrInvalid
+		}
+		return err
 	}
-	return err
+	return nil
+}
+
+func (d *pgDataStore) addHTTP(r *router.Route) error {
+	tx, err := d.pgx.Begin()
+	if err != nil {
+		return err
+	}
+	if err := tx.QueryRow(
+		sqlAddRouteHTTP,
+		r.ParentRef,
+		r.Service,
+		r.Leader,
+		r.Domain,
+		r.Sticky,
+		r.Path,
+	).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt); err != nil {
+		tx.Rollback()
+		return err
+	}
+	if err := d.addCertificateWithTx(tx, r); err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (d *pgDataStore) addTCP(r *router.Route) error {
+	return d.pgx.QueryRow(
+		sqlAddRouteTCP,
+		r.ParentRef,
+		r.Service,
+		r.Leader,
+		r.Port,
+	).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
+}
+
+const sqlSelectCert = `
+SELECT id, created_at, updated_at FROM ` + tableNameCertificates + `
+	WHERE cert_sha256 = $1 AND deleted_at IS NULL`
+
+const sqlAddCert = `
+INSERT INTO ` + tableNameCertificates + ` (cert, key, cert_sha256)
+	VALUES ($1, $2, $3)
+	RETURNING id, created_at, updated_at
+`
+
+const sqlAddCertificate = `
+INSERT INTO ` + tableNameRoutesCertificate + ` (http_route_id, certificate_id)
+	VALUES ($1, $2)
+`
+
+const sqlCleanupCertificates = `
+DELETE FROM ` + tableNameRoutesCertificate + `
+	WHERE http_route_id = $1`
+
+func (d *pgDataStore) AddCert(c *router.Certificate) error {
+	tx, err := d.pgx.Begin()
+	if err != nil {
+		return err
+	}
+	if err := d.addCertWithTx(tx, c); err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (d *pgDataStore) addCertWithTx(tx *pgx.Tx, c *router.Certificate) error {
+	tlsCertSHA256 := hex.EncodeToString(sha256.New().Sum([]byte(c.Cert)))
+	if err := tx.QueryRow(sqlSelectCert, tlsCertSHA256).Scan(&c.ID, &c.CreatedAt, &c.UpdatedAt); err != nil {
+		if err := tx.QueryRow(sqlAddCert, c.Cert, c.Key, tlsCertSHA256).Scan(&c.ID, &c.CreatedAt, &c.UpdatedAt); err != nil {
+			return err
+		}
+	}
+	for _, rid := range c.Routes {
+		if _, err := tx.Exec(sqlCleanupCertificates, rid); err != nil {
+			return err
+		}
+		if _, err := tx.Exec(sqlAddCertificate, rid, c.ID); err != nil && !postgres.IsUniquenessError(err, "") {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *pgDataStore) addCertificateWithTx(tx *pgx.Tx, r *router.Route) error {
```

Comment:
The latter was `addRouteCertWithTx` (casualty of search/replace renaming `router.RouteCert` -> `router.Certificate`).